business/dashboards.go

```diff
@@ -0,0 +1,59 @@
+package business
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/kiali/kiali/kubernetes"
+	"github.com/kiali/kiali/models"
+	"github.com/kiali/kiali/prometheus"
+)
+
+// DashboardsService deals with fetching dashboards from k8s client
+type DashboardsService struct {
+	prom prometheus.ClientInterface
+	mon  *kubernetes.KialiMonitoringClient
+}
+
+// NewDashboardsService initializes this business service
+func NewDashboardsService(mon *kubernetes.KialiMonitoringClient, prom prometheus.ClientInterface) DashboardsService {
+	return DashboardsService{prom: prom, mon: mon}
+}
+
+// GetDashboard returns a dashboard filled-in with target data
+func (in *DashboardsService) GetDashboard(params prometheus.CustomMetricsQuery, template string) (*models.MonitoringDashboard, error) {
+	dashboard, err := in.mon.GetDashboard(params.Namespace, template)
+	if err != nil {
+		return nil, err
+	}
+
+	labels := fmt.Sprintf(`{namespace="%s",app="%s"`, params.Namespace, params.App)
+	if params.Version != "" {
+		labels += fmt.Sprintf(`,version="%s"`, params.Version)
+	}
+	labels += "}"
+	grouping := strings.Join(params.ByLabels, ",")
+
+	wg := sync.WaitGroup{}
+	wg.Add(len(dashboard.Spec.Charts))
+	filledCharts := make([]models.Chart, len(dashboard.Spec.Charts))
+
+	for i, c := range dashboard.Spec.Charts {
+		go func(idx int, chart kubernetes.MonitoringDashboardChart) {
+			defer wg.Done()
+			filledCharts[idx] = models.ConvertChart(chart)
```

Comment:
Well, I know in some places it's done differently. But to me it looks simpler to just call a "converter" than first create an empty struct, then have it fill data from a source. It makes the conversion function fully responsible of building the output object, and I don't have to worry about questions like "is my object correctly initialized before sending to the converter?". If constructors existed in go, that would be a constructor.
Plus (but it's minor detail) I've always seen the "Parse" word being associated to text, maybe I'm biased, but I feel like I'm not "parsing" anything here :)

Reply:
Yes. This is one things I don't like about golang, that hybrid syntax between pure C vs C++ model.
As classes and objects doesn't really exist, the "entry points" some times are confused to read and I can understand different (but all correct) opinions here.

Personally, sometimes I prefer to have a helper function rather than an entry point, specially for minor tasks, but that's pretty biased as well.
But I agree, not on this PR, but code has several inconsistencies that at some point we may want to address.